types/time.go

```diff
@@ -1265,139 +1265,235 @@ func (d Duration) MicroSecond() int {
+	}
+}
+
+func checkHHMMSS(hms [3]int) bool {
+	m, s := hms[1], hms[2]
+	return m < 60 && s < 60
+}
+
+// matchFrac returns overflow, fraction, rest, error
+func matchFrac(str string, fsp int8) (bool, int, string, error) {
+	rest, err := parser.Char(str, '.')
+	if err != nil {
+		return false, 0, str, nil
 	}
 
-	if terror.ErrorEqual(err, io.EOF) {
-		err = ErrTruncatedWrongVal.GenWithStackByArgs("time", origStr)
+	digits, rest, err := parser.Digit(rest, 0)
+	if err != nil {
+		return false, 0, str, err
+	}
+
+	frac, overflow, err := ParseFrac(digits, fsp)
+	if err != nil {
+		return false, 0, str, err
 	}
+
+	return overflow, frac, rest, nil
+}
+
+func matchDuration(str string, fsp int8) (Duration, error) {
+	fsp, err := CheckFsp(int(fsp))
 	if err != nil {
 		return ZeroDuration, errors.Trace(err)
 	}
 
+	if len(str) == 0 {
+		return ZeroDuration, ErrTruncatedWrongVal.GenWithStackByArgs("time", str)
+	}
+
+	negative, rest := isNegativeDuration(str)
+	rest = parser.Space0(rest)
+
+	hhmmss := [3]int{}
+
+	if day, hms, remain, err := matchDayHHMMSS(rest); err == nil {
+		hms[0] += 24 * day
+		rest, hhmmss = remain, hms
+	} else if hms, remain, err := matchHHMMSSDelimited(rest, true); err == nil {
+		rest, hhmmss = remain, hms
+	} else if hms, remain, err := matchHHMMSSCompact(rest); err == nil {
+		rest, hhmmss = remain, hms
+	} else {
+		return ZeroDuration, ErrTruncatedWrongVal.GenWithStackByArgs("time", str)
+	}
+
+	rest = parser.Space0(rest)
+	overflow, frac, rest, err := matchFrac(rest, fsp)
+	if err != nil || len(rest) > 0 {
+		return ZeroDuration, ErrTruncatedWrongVal.GenWithStackByArgs("time", str)
+	}
+
 	if overflow {
-		second++
-		fracPart = 0
+		hhmmssAddOverflow(hhmmss[:], overflow)
+		frac = 0
 	}
-	// Invalid TIME values are converted to '00:00:00'.
-	// See https://dev.mysql.com/doc/refman/5.7/en/time.html
-	if minute >= 60 || second > 60 || (!overflow && second == 60) {
-		return ZeroDuration, ErrTruncatedWrongVal.GenWithStackByArgs("time", origStr)
+
+	if !checkHHMMSS(hhmmss) {
+		return ZeroDuration, ErrTruncatedWrongVal.GenWithStackByArgs("time", str)
 	}
 
-	if day*24+hour > TimeMaxHour {
+	if hhmmss[0] > TimeMaxHour {
 		var t gotime.Duration
-		if sign == -1 {
+		if negative {
 			t = MinTime
 		} else {
 			t = MaxTime
 		}
-		return Duration{Duration: t, Fsp: fsp}, ErrTruncatedWrongVal.GenWithStackByArgs("time", origStr)
+		return Duration{t, fsp}, ErrTruncatedWrongVal.GenWithStackByArgs("time", str)
 	}
 
-	d := gotime.Duration(day*24*3600+hour*3600+minute*60+second)*gotime.Second + gotime.Duration(fracPart)*gotime.Microsecond
-	if sign == -1 {
+	d := gotime.Duration(hhmmss[0]*3600+hhmmss[1]*60+hhmmss[2])*gotime.Second + gotime.Duration(frac)*gotime.Microsecond
+	if negative {
 		d = -d
 	}
-
 	d, err = TruncateOverflowMySQLTime(d)
-	return Duration{Duration: d, Fsp: fsp}, errors.Trace(err)
+	return Duration{d, fsp}, errors.Trace(err)
+}
+
+func canFallbackToDateTime(str string) bool {
```

Comment:
Please add some comments for this function, for example in what cases we can fallback to `DateTime`.

Reply:
Done.